pkg/gcrane/list.go

```diff
@@ -29,50 +27,67 @@ import (
 func init() { Root.AddCommand(NewCmdList()) }
 
 func NewCmdList() *cobra.Command {
-	return &cobra.Command{
+	recursive := false
+	cmd := &cobra.Command{
 		Use:   "ls",
 		Short: "List the contents of a repo",
 		Args:  cobra.ExactArgs(1),
-		Run:   ls,
+		Run: func(_ *cobra.Command, args []string) {
+			ls(args[0], recursive)
+		},
 	}
+
+	cmd.Flags().BoolVarP(&recursive, "recursive", "r", false, "Whether to recurse through repos")
+
+	return cmd
 }
 
-func ls(_ *cobra.Command, args []string) {
-	r := args[0]
-	repo, err := name.NewRepository(r, name.WeakValidation)
+func ls(root string, recursive bool) {
+	repo, err := name.NewRepository(root, name.WeakValidation)
 	if err != nil {
 		log.Fatalln(err)
 	}
-	auth, err := authn.DefaultKeychain.Resolve(repo.Registry)
-	if err != nil {
-		log.Fatalln(err)
+
+	auth := google.WithAuthFromKeychain(authn.DefaultKeychain)
+
+	if recursive {
+		if err := google.Walk(repo, printImages, auth); err != nil {
+			log.Fatalln(err)
+		}
 	}
-	tags, err := google.List(repo, auth, http.DefaultTransport)
+
+	tags, err := google.List(repo, auth)
 	if err != nil {
 		log.Fatalln(err)
 	}
 
-	// Track what we saw in the response so we can fall back to non-GCR behavior.
-	gcrSpecific := false
+	if len(tags.Manifests) == 0 && len(tags.Children) == 0 {
+		// If we didn't see any GCR extensions, just list the tags like normal.
+		for _, tag := range tags.Tags {
+			fmt.Printf("%s:%s\n", repo, tag)
+		}
+		return
+	}
 
 	for _, child := range tags.Children {
 		fmt.Printf("%s/%s\n", repo, child)
 	}
 
+	_ = printImages(repo, tags, err)
```

Comment:
Can you add a comment about why this error is unhandled?

Reply:
Handled it instead.